File: Semaphore_Examples_PY/barrier.py
import threading

class Barrier :
    def __init__ (self, n):
        self.n = n
        self.count = 0
        self.mutex = threading.Semaphore(1)
        self.turnstile = threading.Semaphore(0)
        self.turnstile2 = threading.Semaphore(0)

    def phase1 (self):
        self.mutex.acquire()
        self.count += 1
        if self.count == self.n:
            self.turnstile.release(self.n)
        self.mutex.release()
        self.turnstile.acquire()

    def phase2 (self):
        self.mutex.acquire()
        self.count -= 1
        if self.count == 0:
            self.turnstile2.release(self.n)
        self.mutex.release()
        self.turnstile2.acquire()

    def wait (self):
        self.phase1()
        self.phase2()


barrier = Barrier(3)
barrier.wait()

File: Semaphore_Examples_CPP/Barrier.cpp
//
// Created by ozgur on 7/21/2024.
//

#include "Barrier.h"

Barrier::Barrier(int _n) :
        n(_n),
        count(0),
        mutex(1),
        turnstile(0),
        turnstile2(0)
{}

void Barrier::phase1()
{
    mutex.acquire(); // wait , lock
    count++; // counts the thread which gets in
    if (count == n) { turnstile.release(n); } // Allowe to pass whole threads
    mutex.release();

    turnstile.acquire();
}

void Barrier::phase2()
{
    mutex.acquire(); // wait , lock
    count--; // counts the thread which gets out
    if (count == 0) { turnstile2.release(n); } // Allowe to pass whole threads
    mutex.release();

    turnstile2.acquire();
}

void Barrier::wait()
{
    phase1();
    phase2();
}